tiri/src/lib.rs

```diff
@@ -50,35 +154,186 @@ impl<'t> Interp<'t> {
     }
 
     /// Start interpreting the trace.
-    pub fn run(&self) {
-        let state = InterpState::new();
+    pub fn run(&self, mut state: InterpState) -> InterpState {
+        let trace_len = self.trace.len();
 
         // The main interpreter loop.
-        loop {
-            let op = self.trace.op(state.trace_pos);
+        while !state.abort && state.pc < trace_len {
+            let op = self.trace.op(state.pc);
             match op {
-                TirOp::Statement(st) => self.interp_stmt(st),
-                TirOp::Guard(g) => self.interp_guard(g),
-            }
+                TirOp::Statement(stmt) => self.interp_stmt(&mut state, stmt),
+                TirOp::Guard(grd) => self.interp_guard(&mut state, grd),
+            };
         }
+
+        state
     }
 
     /// Interpret the specified statement.
-    fn interp_stmt(&self, _stmt: &Statement) {
-        unimplemented!();
+    fn interp_stmt(&self, state: &mut InterpState, stmt: &Statement) {
+        match stmt {
+            Statement::Assign(plc, rval) => match (&plc.base, &plc.projections.len()) {
+                (PlaceBase::Local(l), 0) => {
+                    state.store_local(&l, self.eval_rvalue(state, rval).clone());
+                    state.pc += 1;
+                }
+                _ => unimplemented!("unhandled assignment"),
+            },
+            _ => panic!("unhandled statement: {}", stmt),
+        }
+    }
+
+    fn eval_rvalue(&self, state: &InterpState, rval: &Rvalue) -> Value {
+        match rval {
+            Rvalue::Use(o) => self.eval_operand(state, o),
+            Rvalue::BinaryOp(oper, o1, o2) => self.eval_binop(state, oper, o1, o2, false),
+            Rvalue::CheckedBinaryOp(oper, o1, o2) => self.eval_binop(state, oper, o1, o2, true),
+            _ => panic!("unimplemented rvalue eval"),
+        }
+    }
+
+    fn eval_place(&self, state: &InterpState, place: &Place) -> Value {
```

Comment:
What does the `place` in `eval_place` mean?

Reply:
> (Although it's defined in Miri, the same type is directly re-used in MIR).

Hold up, that's nonsense! I meant to link to this one:
https://github.com/softdevteam/ykrustc/blob/e103291e3553ec58444caa056765e5c3312caae4/src/librustc/mir/mod.rs#L1841

(There are two `Place`s in upstream rust, one for MIR, once for Miri).